test/e2e/framework/apihelper/helper.go

```diff
@@ -0,0 +1,370 @@
+	if h.unsupportedPlatform {
+		framework.Logf("This test runs with an API server that does not report endpoints to the kubernetes service in the default namespace." + " " +
+			"We cannot determine how many control plane API servers are running, which means we can't verify that all instances have stopped serving."  + " " +
+			"You can ignore this warning if you are running only a single API server instance.")
+		return nil
+	}
+	return wait.Poll(500*time.Millisecond, waitTimeout, mustSucceedForAllAPIServers(h.clients, conditionFn))
+}
+
+// CleanUp executed the cleanup routine.
+// This method MUST be called after a successful call to NewAPIEndpointHelper function.
+func (h *APIHelper) CleanUp() {
+	h.cleanUpFn()
+}
+
+// NewAPIEndpointHelper is a convenience function for checking conditions against all deployed Kube API servers.
+// Use WaitForAll for checking a condition on all servers and CleanUp for cleaning up after the test.
+// Note that at the moment we only support detecting servers by getting API endpoints appropriately.
+func NewAPIEndpointHelper(client clientset.Interface, namespace string) (*APIHelper, error) {
+	apiHelper := &APIHelper{
+		cleanUpFn: func(){/* noop */},
+	}
+	kubeAPIServersPorts, cleanUpFn, err := setupAPIServersProxyPodAndPortForward(client, namespace)
+	if err != nil {
+		cleanUpFn()
+		if err == unsupportedPlatformErr {
+			// suppress the error, we are going to warn the caller when it calls WaitForAll method
+			apiHelper.unsupportedPlatform = true
+			return apiHelper, nil
+		}
+		return nil, err
+	}
+	config, err := framework.LoadConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	apiHelper.cleanUpFn = cleanUpFn
+	for _, kubeAPIServerPort := range kubeAPIServersPorts {
+		configCopy := *config
+		setDefaultServerName(&configCopy)
+		configCopy.Host = fmt.Sprintf("https://localhost:%d", kubeAPIServerPort)
+		newClient, err := clientset.NewForConfig(&configCopy)
+		if err != nil {
+			return nil, err
+		}
+		apiHelper.clients = append(apiHelper.clients, newClient)
+	}
+	return apiHelper, nil
+}
+
+// mustSucceedForAllAPIServers calls f multiple times on success and only returns true if all calls are successful.
+// This is necessary to avoid flaking tests where one call might hit a good apiserver while in HA other apiservers
+// might be lagging behind.
+func mustSucceedForAllAPIServers(kubeClients []*clientset.Clientset, f func(*clientset.Clientset) (bool, error)) func() (bool, error) {
+	return func() (bool, error) {
+		for i := 0; i < len(kubeClients); i++ {
+			ok, err := f(kubeClients[i])
+			if err != nil || !ok {
+				return ok, err
+			}
+		}
+		return true, nil
+	}
+}
+
+// setupAPIServersProxyPodAndPortForward a convenience method that creates and runs a pod that proxies connections to the API servers.
+// It also uses kubectl port-forward to route local connections to that pod.
+func setupAPIServersProxyPodAndPortForward(client clientset.Interface, namespace string) ([]int, func(), error) {
```

Comment:
we have a nice dialer that does this, it was added recently https://github.com/kubernetes/kubernetes/pull/102050/commits/5e9076da93c8bd8d51dd9c6d2d4f87d5a269e586#diff-1603c90cc2b55449a34718b2333960968302a0e5d8c9d43cafeca2cae8c0f205